<?xml version="1.0"?>
<robot name="ai_car">
    <material name="grey">
        <color rgba="0.5 0.5 0.5 1.0"/>
    </material>

    <link name="base_footprint"/>

    <joint name="base_joint" type="fixed">
        <parent link="base_footprint"/>
        <child link="base_link"/>
        <origin xyz="0.0 0.0 0.010" rpy="0 0 0"/>
    </joint>

    <link name="base_link">
        <visual>
            <geometry>
                <mesh filename="package://ai_car/meshs/base_link.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0.0 0 0.09" rpy="0 0 0"/>
            <geometry>
                <box size="0.27 0.14 0.19"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="8.2573504e-01"/>
            <inertia ixx="2.2124416e-03" ixy="-1.2294101e-05" ixz="3.4938785e-05"
                    iyy="2.1193702e-03" iyz="-5.0120904e-06"
                    izz="2.0064271e-03" />
        </inertial>
    </link>

    <joint name="wheel_left_joint" type="continuous">
        <parent link="base_link"/>
        <child link="wheel_left_link"/>
        <axis xyz="0 0 1"/>
        <origin xyz="0.092 0.08 0.023" rpy="-1.57 0 0"/>
    </joint>

    <link name="wheel_left_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="1.57 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/left_tire.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.018" radius="0.033"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0" />
            <mass value="2.8498940e-02" />
            <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
                    iyy="1.1192413e-05" iyz="-1.4400107e-11"
                    izz="2.0712558e-05" />
        </inertial>
    </link>

    <joint name="wheel_right_joint" type="continuous">
        <parent link="base_link"/>
        <child link="wheel_right_link"/>
        <axis xyz="0 0 1"/>
        <origin xyz="0.092 -0.08 0.023" rpy="-1.57 0 0"/>
    </joint>

    <link name="wheel_right_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="1.57 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/right_tire.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.018" radius="0.033"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0" />
            <mass value="2.8498940e-02" />
            <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
                    iyy="1.1192413e-05" iyz="-1.4400107e-11"
                    izz="2.0712558e-05" />
        </inertial>
    </link>

    <joint name="wheel_rear_left_joint" type="continuous">
        <parent link="base_link"/>
        <child link="wheel_rear_left_link"/>
        <axis xyz="0 0 1"/>
        <origin xyz="-0.092 0.08 0.023" rpy="-1.57 0 0"/>
    </joint>

    <link name="wheel_rear_left_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="1.57 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/left_tire.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.018" radius="0.033"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0" />
            <mass value="2.8498940e-02" />
            <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
                    iyy="1.1192413e-05" iyz="-1.4400107e-11"
                    izz="2.0712558e-05" />
        </inertial>
    </link>

    <joint name="wheel_rear_right_joint" type="continuous">
        <parent link="base_link"/>
        <child link="wheel_rear_right_link"/>
        <axis xyz="0 0 1"/>
        <origin xyz="-0.092 -0.08 0.023" rpy="-1.57 0 0"/>
    </joint>

    <link name="wheel_rear_right_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="1.57 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/right_tire.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.018" radius="0.033"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0" />
            <mass value="2.8498940e-02" />
            <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
                    iyy="1.1192413e-05" iyz="-1.4400107e-11"
                    izz="2.0712558e-05" />
        </inertial>
    </link>

    <joint name="scan_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_scan"/>
        <origin xyz="-0.064 0.0 0.22" rpy="0.0 0.0 0.0"/>
    </joint>

    <link name="base_scan">
        <visual>
            <origin xyz="0.0 0.0 -0.022" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/RP-lidar-A1.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 -0.0065" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.0315" radius="0.055"/>
            </geometry>
        </collision>

        <inertial>
            <mass value="0.114" />
            <origin xyz="0 0 0" />
            <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                    iyy="0.001" iyz="0.0"
                    izz="0.001" />
        </inertial>
    </link>

    <joint name="imu_joint" type="fixed">
        <parent link="base_link"/>
        <child link="imu_link"/>
        <origin xyz="0.04 0.0 0.0" rpy="0.0 0.0 0.0"/>
    </joint>

    <link name="imu_link"/>
    
    <joint name="arm_joint_1" type="revolute">
        <parent link="base_link"/>
        <child link="arm_link_1"/>
        <axis xyz="0 1 0"/>
        <limit lower="0" upper="1.57" velocity="50.0" effort="1000.0"/>
        <origin xyz="0.1735 -0.020375 0.03225" rpy="-3.14 1.57 0.0"/>
    </joint>

    <link name="arm_link_1">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="3.14 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/arm_link_1.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="3.14 0 0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/arm_link_1.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="3.0876154e-04 0.0000000e+00 -1.2176461e-04" />
            <mass value="7.9119962e-02" />
            <inertia ixx="1.2505234e-05" ixy="0.0" ixz="-1.7855208e-07"
                    iyy="2.1898364e-05" iyz="0.0"
                    izz="1.9267361e-05" />
        </inertial>
    </link>
    
    <joint name="arm_joint_2" type="revolute">
        <parent link="arm_link_1"/>
        <child link="arm_link_2"/>
        <axis xyz="0 1 0"/>
        <limit lower="-1.57" upper="0" velocity="50.0" effort="1000.0"/>
        <origin xyz="-0.072 0.0 -0.0568" rpy="0 1.57 0"/>

    </joint>

    <link name="arm_link_2">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="3.14 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/arm_link_2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="3.14 -0.785 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/arm_link_2.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="3.0876154e-04 0.0000000e+00 -1.2176461e-04" />
            <mass value="7.9119962e-02" />
            <inertia ixx="1.2505234e-05" ixy="0.0" ixz="-1.7855208e-07"
                    iyy="2.1898364e-05" iyz="0.0"
                    izz="1.9267361e-05" />
        </inertial>
    </link>
    
    <joint name="tool_joint" type="revolute">
        <parent link="arm_link_2"/>
        <child link="arm_tool"/>
        <axis xyz="0 0 1"/>
        <limit lower="-0.4" upper="0.8" velocity="50.0" effort="1000.0"/>
        <origin xyz="0.09125 -0.00625 0.021" rpy="0 0 0"/>
    </joint>

    <link name="arm_tool">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="3.14 0.0 0.0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/arm_tool.stl" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="grey"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="3.14 0 0"/>
            <geometry>
                <mesh filename="package://ai_car/meshs/arm_tool.stl" scale="0.001 0.001 0.001"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="3.0876154e-04 0.0000000e+00 -1.2176461e-04" />
            <mass value="7.9119962e-02" />
            <inertia ixx="1.2505234e-05" ixy="0.0" ixz="-1.7855208e-07"
                    iyy="2.1898364e-05" iyz="0.0"
                    izz="1.9267361e-05" />
        </inertial>
    </link>

</robot>